### PR TITLE
scale a cluster using REST API

### PR DESCRIPTION
**Pull request by xiancao**
Add the test infrastructure support for scaling a cluster using REST API.

Jira task: https://jira.oraclecorp.com/jira/browse/OWLS-82419

Jenkins result: https://build.weblogick8s.org:8443/view/all/job/weblogic-kubernetes-operator-kind-new/279/
The test oracle.weblogic.kubernetes.ItSessionMigration.testSessionMigration failed on develop branch too. See https://build.weblogick8s.org:8443/view/all/job/weblogic-kubernetes-operator-kind-new/268/
https://build.weblogick8s.org:8443/view/all/job/weblogic-kubernetes-operator-model-in-image-tests-10/266/

### CODE REVIEW

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/actions/impl/Domain.java [hunk 1]
```diff
@@ -173,4 +177,63 @@ public static boolean scaleCluster(String domainUid, String namespace, String cl
...
+   * @return true if REST call succeeds, false otherwise
+   */
+  public static boolean scaleClusterWithRestApi(String domainUid,
+                                                String clusterName,
+                                                int numOfServers,
+                                                int externalRestHttpsPort,
+                                                String opNamespace,
+                                                String opServiceAccount) {
+
+    String secretName = Secret.getSecretFromServiceAccount(opNamespace, opServiceAccount);
```
[the remark this answers is not in the document]
Comment (vanajamukkara) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/actions/impl/Domain.java:198]: add this comment 
//Get the service account token stored in secret to authenticate as the service account.

[the remark this answers is not in the document]
Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/actions/impl/Domain.java:198]: fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/actions/impl/Domain.java [hunk 2]
```diff
@@ -173,4 +177,63 @@ public static boolean scaleCluster(String domainUid, String namespace, String cl
...
+    logger.info("Got secret {0} in operator namespace {1}", secretName, opNamespace);
```
[the remark this answers is not in the document]
Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/actions/impl/Domain.java:199]: fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/actions/impl/Domain.java [hunk 3]
```diff
@@ -173,4 +177,63 @@ public static boolean scaleCluster(String domainUid, String namespace, String cl
...
+
+    String secretToken = Secret.getSecretEncodedToken(opNamespace, secretName);
```
[the remark this answers is not in the document]
Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/actions/impl/Domain.java:201]: fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonTestUtils.java [hunk 4]
```diff
@@ -741,6 +775,42 @@ public static void scaleAndVerifyCluster(String clusterName,
                                            String curlCmd,
                                            List<String> expectedServerNames) {
 
+    scaleAndVerifyCluster(clusterName, domainUid, domainNamespace, manageServerPodNamePrefix, replicasBeforeScale,
+        replicasAfterScale, false, 0, "", "", curlCmd, expectedServerNames);
+  }
+
+  /** Scale the WebLogic cluster to specified number of servers.
```
[the remark this answers is not in the document]
Comment (xiancao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonTestUtils.java:782]: fixed

**Comment by vanajamukkara** [comment]
@xiancao tests are failing on https://build.weblogick8s.org:8443/view/all/job/weblogic-kubernetes-operator-model-in-image-tests-10/262/consoleFull

```
<06-04-2020 10:10:12> <INFO> <oracle.weblogic.kubernetes.extensions.IntegrationTestWatcher printHeader> <
-----------------------------------------------------------------------
Starting beforeAll for oracle.weblogic.kubernetes.ItScaleMiiDomainNginx
-----------------------------------------------------------------------
>
<06-04-2020 10:10:12> <INFO> <oracle.weblogic.kubernetes.extensions.IntegrationTestWatcher resolveParameter> <Created a new namespace called ns-khcq>
<06-04-2020 10:10:12> <INFO> <oracle.weblogic.kubernetes.extensions.IntegrationTestWatcher resolveParameter> <Created a new namespace called ns-tdwm>
<06-04-2020 10:10:12> <INFO> <oracle.weblogic.kubernetes.extensions.IntegrationTestWatcher resolveParameter> <Created a new namespace called ns-rceo>
<06-04-2020 10:10:12> <INFO> <oracle.weblogic.kubernetes.extensions.IntegrationTestWatcher resolveParameter> <[ns-khcq, ns-tdwm, ns-rceo]>
<06-04-2020 10:10:12> <INFO> <oracle.weblogic.kubernetes.ItScaleMiiDomainNginx initAll> <Get a unique namespace for operator>
<06-04-2020 10:10:12> <INFO> <oracle.weblogic.kubernetes.ItScaleMiiDomainNginx initAll> <Get a unique namespace for WebLogic domain>
<06-04-2020 10:10:12> <INFO> <oracle.weblogic.kubernetes.ItScaleMiiDomainNginx initAll> <Get a unique namespace for NGINX>
<06-04-2020 10:10:12> <INFO> <oracle.weblogic.kubernetes.utils.TestUtils getNextFreePort> <next free port is: 31,001>
<06-04-2020 10:10:12> <INFO> <oracle.weblogic.kubernetes.utils.CommonTestUtils installAndVerifyOperator> <Creating service account>
<06-04-2020 10:10:12> <INFO> <oracle.weblogic.kubernetes.utils.CommonTestUtils installAndVerifyOperator> <Created service account: ns-khcq-sa>
<06-04-2020 10:10:12> <INFO> <oracle.weblogic.kubernetes.utils.CommonTestUtils installAndVerifyOperator> <operator image name ****/weblogick8s/oracle/weblogic-kubernetes-operator:xc-restscale262>
<06-04-2020 10:10:12> <INFO> <oracle.weblogic.kubernetes.utils.CommonTestUtils installAndVerifyOperator> <Creating Docker registry secret in namespace ns-khcq>
<06-04-2020 10:10:13> <INFO> <oracle.weblogic.kubernetes.actions.impl.primitive.Command execute> <Executing command /scratch/jenkins/workspace/weblogic-kubernetes-operator-model-in-image-tests-10/new-integration-tests/../kubernetes/samples/scripts/rest/generate-external-rest-identity.sh -a "DNS:130.61.49.225,DNS:localhost,IP:127.0.0.1" -n ns-khcq -s weblogic-operator-external-rest-identity>
keytool error: java.lang.RuntimeException: java.io.IOException: DNSName components must begin with a letter
<06-04-2020 10:10:13> <SEVERE> <oracle.weblogic.kubernetes.actions.impl.primitive.Command execute> <The command execution failed because it returned non-zero exit value: ExecResult: exitValue = 1, stdout = keytool error: java.lang.RuntimeException: java.io.IOException: DNSName components must begin with a letter, stderr = .>
<06-04-2020 10:10:13> <INFO> <oracle.weblogic.kubernetes.extensions.IntegrationTestWatcher printHeader> <
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
BeforeAll failed oracle.weblogic.kubernetes.ItScaleMiiDomainNginx
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
>
```

**Comment by xiancao** [comment]
@vanajamukkara Thanks for pointing it out. I just saw no test failure. I did not notice the test was failing in BeforeAll.

**Comment by xiancao** [comment]
The K8S_NODEPORT_HOST returns IP address instead of DNS names. The script runs fine in kind cluster but failed in cluster 10.

**Comment by vanajamukkara** [comment]
This is how we do in old tests 
```
    // here we are assuming that if the "host name" starts with a digit, then it is actually
    // an IP address, and so we need to use the "IP" prefix in the SANS.
    if (Character.isDigit(TestUtils.getHostName().charAt(0))) {
      sb.append(" IP:");
    } else {
      sb.append(" DNS:");
    }
```

**Pull request merged**